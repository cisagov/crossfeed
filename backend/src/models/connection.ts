import { createConnection, Connection } from 'typeorm';
import {
  Domain,
  Service,
  Vulnerability,
  Scan,
  Organization,
  User,
  Role,
  ScanTask
} from '.';

let connection: Connection | null = null;

const connectDb = async (logging?: boolean) => {
  const connection = createConnection({
    type: 'postgres',
    host: process.env.DB_HOST,
    port: parseInt(process.env.DB_PORT ?? ''),
    username: process.env.DB_USERNAME,
    password: process.env.DB_PASSWORD,
    database: process.env.DB_NAME,
    entities: [
      Domain,
      Service,
      Vulnerability,
      Scan,
      Organization,
      User,
      Role,
      ScanTask
    ],
    synchronize: false,
    name: 'default',
    dropSchema: false,
    logging: logging ?? false
  });
  return connection;
};

export const connectToDatabase = async (logging?: boolean) => {
  if (!connection?.isConnected) {
    connection = await connectDb(logging);
  }
  if (typeof jest === 'undefined') {
    console.log('=> DB Connected');
  }
  return connection;
};
