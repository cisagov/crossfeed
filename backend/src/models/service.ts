import {
  Entity,
  Column,
  Index,
  PrimaryGeneratedColumn,
  ManyToOne,
  BaseEntity,
  CreateDateColumn
} from 'typeorm';
import { Domain } from './domain';
import { Scan } from './scan';

@Entity()
@Index(['port', 'domain'], { unique: true })
export class Service extends BaseEntity {
  @PrimaryGeneratedColumn('uuid')
  id: string;

  @CreateDateColumn()
  createdAt: Date;

  @ManyToOne((type) => Domain, (domain) => domain.services, {
    onDelete: 'CASCADE',
    onUpdate: 'CASCADE'
  })
  domain: Domain;

  @ManyToOne((type) => Scan, {
    onDelete: 'CASCADE',
    onUpdate: 'CASCADE'
  })
  discoveredBy: Scan;

  @Column()
  port: number;

  @Column({
    nullable: true,
    type: 'varchar'
  })
  service: string | null;

  @Column({
    nullable: true,
    type: 'timestamp'
  })
  lastSeen: Date | null;

  @Column({
    nullable: true,
    type: 'text'
  })
  banner: string | null;

  /** Censys Metadata */
  @Column({
    type: 'jsonb',
    default: {}
  })
  censysMetadata: {
    product: string;
    revision: string;
    description: string;
    version: string;
    manufacturer: string;
  } | null;

  /** Censys Ipv4 results */
  @Column({
    type: 'jsonb',
    default: {}
  })
  censysIpv4Results: {
    [x: string]: any;
  };
}
