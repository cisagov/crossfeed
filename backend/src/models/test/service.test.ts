import { connectToDatabase } from '../connection';
import { Service } from '../service';

describe('service', () => {
  beforeAll(async () => {
    await connectToDatabase();
  });
  test('set products', async () => {
    const service = await Service.create({
      port: 443
    }).save();
    service.wappalyzerResults = [
      {
        cpe: 'cpe1',
        name: 'name',
        version: 'version',
        slug: 'slug',
        icon: 'icon',
        website: 'website',
        confidence: 0,
        categories: []
      }
    ];
    await service.save();
    expect(service.products).toMatchSnapshot();
  });
  test('set products should not include a blacklisted product', async () => {
    const service = await Service.create({
      port: 443
    }).save();
    service.wappalyzerResults = [
      {
        cpe: 'cpe:/a:apache:tomcat',
        name: 'name',
        version: '1.1',
        slug: 'slug',
        icon: 'icon',
        website: 'website',
        confidence: 0,
        categories: []
      }
    ];
    await service.save();
    expect(service.products.length).toEqual(0);
  });
});
