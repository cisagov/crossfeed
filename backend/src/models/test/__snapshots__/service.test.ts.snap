// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`service set products 1`] = `
Array [
  Object {
    "cpe": "cpe1",
    "icon": "icon",
    "name": "name",
    "tags": Array [],
    "version": "version",
  },
]
`;
